### PR TITLE
[ELF] Reject error-prone meta characters in input section description

### PR DESCRIPTION
**Pull request by MaskRay**
The lexer is overly permissive. When parsing file patterns in an input
section description and there is a missing `)`, we would accept many
non-sensible tokens (e.g. `}`) as patterns, leading to confusion, e.g.
`*(SORT_BY_ALIGNMENT(SORT_BY_NAME(.text*)) } PROVIDE_HIDDEN(__code_end = .)`
(#81804).

Ideally, the lexer should be stateful to report more errors like GNU ld
and get rid of hacks like `ScriptLexer::maybeSplitExpr`, but that would
require a large rewrite of the lexer. For now, just reject certain
non-wildcard meta characters to detect common mistakes.

### CODE REVIEW

File: lld/test/ELF/linkerscript/wildcards.s [hunk 1]
```diff
@@ -91,24 +91,31 @@ SECTIONS {
...
 }
 
-## Test a few non-wildcard meta characters rejected by GNU ld.
+## Test a few non-wildcard characters rejected by GNU ld.
 
 #--- lbrace.lds
-# RUN: ld.lld -T lbrace.lds a.o -o out
+# RUN: not ld.lld -T lbrace.lds a.o 2>&1 | FileCheck %s --check-prefix=ERR-LBRACE --match-full-lines --strict-whitespace
+#      ERR-LBRACE:{{.*}}: section pattern is expected
+# ERR-LBRACE-NEXT:>>>   .text : { *(.a* { ) }
```
Comment (smithp35) [inline lld/test/ELF/linkerscript/wildcards.s:99]: Is it worth adding a case when there is no space between the disallowed character for example `.text : { *(.a*{)` as I understand it (,),{,} are lexed as a single token so the spaces shouldn't matter. 

From reading the line and the tests all having the character separated by spaces, it made me double check that we could catch more than just a single isolated character.
```
if (s.size() == 1 && strchr("(){}", s[0]))
```

File: lld/ELF/ScriptParser.cpp [hunk 2]
```diff
@@ -717,9 +717,19 @@ SmallVector<SectionPattern, 0> ScriptParser::readInputSectionsList() {
 
     StringMatcher SectionMatcher;
     // Break if the next token is ), EXCLUDE_FILE, or SORT*.
-    while (!errorCount() && peek() != ")" && peek() != "EXCLUDE_FILE" &&
-           peekSortKind() == SortSectionPolicy::Default)
+    while (!errorCount() && peekSortKind() == SortSectionPolicy::Default) {
+      StringRef s = peek();
+      if (s == ")" || s == "EXCLUDE_FILE")
+        break;
+      // Detect common mistakes that certain non-wildcard meta characters used
```
Comment (smithp35) [inline lld/ELF/ScriptParser.cpp:724]: Suggest
// Detect common mistakes when certain non-wildcard meta characters are used without a closing `)`

**Review by smithp35 — COMMENTED** [review]
A few small suggestions on the test and a comment.

File: lld/test/ELF/linkerscript/wildcards.s (hunk 1 above)
Comment (MaskRay) [reply lld/test/ELF/linkerscript/wildcards.s:99]: Thanks for the suggestion. Updated. Changed `s.size() == 1` to `!s.empty()` to be clearer that we just guard again `""` special case.

File: lld/ELF/ScriptParser.cpp (hunk 2 above)
Comment (MaskRay) [reply lld/ELF/ScriptParser.cpp]: Updated

**Review by smithp35 — APPROVED** [review]
LGTM, thanks for the updates.

**Review by mysterymath — APPROVED** [review]
LGTM, thanks!

**Pull request merged**